dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.cpp

```diff
@@ -0,0 +1,409 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Storages/DeltaMerge/LocalIndexerScheduler.h>
+#include <common/logger_useful.h>
+
+namespace DB::DM
+{
+
+LocalIndexerScheduler::LocalIndexerScheduler(const Options & options)
+    : logger(Logger::get())
+    , pool(std::make_unique<ThreadPool>(options.pool_size, options.pool_size, options.pool_size + 1))
+    , pool_max_memory_limit(options.memory_limit)
+    , pool_current_memory(0)
+{
+    // QueueSize = PoolSize+1, because our scheduler will try to schedule next task
+    // right before the current task is finished.
+
+    LOG_INFO(
+        logger,
+        "Initialized LocalIndexerScheduler, pool_size={}, memory_limit_mb={:.1f}",
+        options.pool_size,
+        static_cast<double>(options.memory_limit) / 1024 / 1024);
+
+    if (options.auto_start)
+        start();
+}
+
+LocalIndexerScheduler::~LocalIndexerScheduler()
+{
+    LOG_INFO(logger, "LocalIndexerScheduler is destroying. Waiting scheduler and tasks to finish...");
+
+    // First quit the scheduler. Don't schedule more tasks.
+    is_shutting_down = true;
+    {
+        std::unique_lock lock(mutex);
+        scheduler_need_wakeup = true;
+        scheduler_notifier.notify_all();
+    }
+
+    if (is_started)
+        scheduler_thread.join();
+
+    // Then wait all running tasks to finish.
+    pool.reset();
+
+    LOG_INFO(logger, "LocalIndexerScheduler is destroyed");
+}
+
+void LocalIndexerScheduler::start()
+{
+    if (is_started)
+        return;
+
+    scheduler_thread = std::thread([this]() { schedulerLoop(); });
+    is_started = true;
+}
+
+void LocalIndexerScheduler::waitForFinish()
+{
+    while (true)
+    {
+        std::unique_lock lock(mutex);
+        if (all_tasks_count == 0 && running_tasks_count == 0)
+            return;
+        on_finish_notifier.wait(lock);
+    }
+}
+
+void LocalIndexerScheduler::pushTask(const Task & task)
+{
+    if (pool_max_memory_limit > 0 && task.request_memory > pool_max_memory_limit)
```

Comment:
Should we better record the total memory requested by all tasks?

Reply:
Pending tasks are not running, will not consume memory.

Reply:
OK, so if a task requires more than pool_max_memory_limit, it can never be scheduled.